src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,617 @@
+ *                              .addSalt(salt).thenExpand(info, 32);
+ *
+ *    kdfHkdf.deriveKey("AES", kdfParameterSpec);
+ *}
+ *
+ * @see KDFParameters
+ * @see SecretKey
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public final class KDF {
+    private static final Debug debug = Debug.getInstance("jca", "KDF");
+
+    private static final Debug pdebug = Debug.getInstance("provider",
+                                                          "Provider");
+    private static final boolean skipDebug = Debug.isOn("engine=")
+                                             && !Debug.isOn("kdf");
+
+    // The provider
+    private Provider provider;
+
+    // The provider implementation (delegate)
+    private KDFSpi spi;
+
+    // The name of the KDF algorithm.
+    private final String algorithm;
+
+    // Additional KDF configuration parameters
+    private final KDFParameters kdfParameters;
+
+    // next service to try in provider selection
+    // null once provider is selected
+    private Service firstService;
+
+    // remaining services to try in provider selection
+    // null once provider is selected
+    private Iterator<Service> serviceIterator;
+
+    private final Object lock;
+
+    int DERIVE_KEY = 0;
+    int DERIVE_DATA = 1;
+
+    /**
+     * Instantiates a KDF object.
+     *
+     * @param keyDerivSpi
+     *     the delegate
+     * @param provider
+     *     the provider
+     * @param algorithm
+     *     the algorithm
+     * @param kdfParameters
+     *     the algorithm parameters
+     */
+    private KDF(KDFSpi keyDerivSpi, Provider provider, String algorithm,
+                KDFParameters kdfParameters) {
+        this.spi = keyDerivSpi;
+        this.provider = provider;
+        this.algorithm = algorithm;
+        this.kdfParameters = kdfParameters;
+        firstService = null;
+        serviceIterator = null;
+        lock = new Object();
+    }
+
+    private KDF(Service s, Iterator<Service> t, String algorithm,
+                KDFParameters kdfParameters) {
+        firstService = s;
+        serviceIterator = t;
+        this.algorithm = algorithm;
+        this.kdfParameters = kdfParameters;
+        lock = new Object();
+    }
+
+    /**
+     * Returns the algorithm name of this {@code KDF} object.
+     *
+     * @return the algorithm name of this {@code KDF} object
+     */
+    public String getAlgorithm() {
+        return this.algorithm;
+    }
+
+    /**
+     * Returns the name of the provider.
+     *
+     * @return the name of the provider
+     */
+    public String getProviderName() {
+        chooseFirstProvider();
+        return provider.getName();
+    }
+
+    /**
+     * Returns the {@code KDFParameters} used to initialize the object.
+     *
+     * @return the parameters used to initialize the object; may be {@code null}
+     * if no additional parameters were provided
+     */
+    public KDFParameters getKDFParameters() {
```

Comment:
I still want to know if this method always returns null if only getInstance(alg) is called without params. Or, when there are default params, they will be returned.

Reply:
This will imp[act the API, e.g. whether KDFSpi needs an `engineGetKDFParameters()` method. BTW, I find the name 
"KDFParameters" very confusing as there is a parameter spec in the `deriveKey()` method named "kdfParameterSpec". The JCA convention uses `AlgorithmParameters` and `AlgorithmParameterSpec` interchangeably, the former is an opaque representation vs the later is a transparent form which algorithm specific values are readily retrievable. 
Maybe name the parameters more specifically based on their purpose? Say, "configParams" for the ones used in `getInstance(...)`, "deriveParams" for the one in `deriveKey()`.